presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionDescriptor.java

```diff
@@ -0,0 +1,170 @@
+            boolean isPruningSubfields,
+            List<LambdaDescriptor> lambdaDescriptors,
+            Set<Integer> callArgumentsToPassOuterLambdaSubfields,
+            Function<Set<Subfield>, Optional<Set<Subfield>>> outerLambdaSubfieldsTransform)
+    {
+        this.isAccessingSubfieldsInternally = isAccessingSubfieldsInternally;
+        this.isTransformingInput = isTransformingInput;
+        this.isPruningSubfields = isPruningSubfields;
+        this.lambdaDescriptors = unmodifiableList(requireNonNull(lambdaDescriptors, "lambdaDescriptors is null"));
+        this.callArgumentsToPassOuterLambdaSubfields = callArgumentsToPassOuterLambdaSubfields != null ? unmodifiableSet(callArgumentsToPassOuterLambdaSubfields) : null;
+        this.outerLambdaSubfieldsTransform = outerLambdaSubfieldsTransform;
+    }
+
+    public static FunctionDescriptor defaultFunctionDescriptor()
+    {
+        return DEFAULT;
+    }
+
+    public boolean isAccessingSubfieldsInternally()
+    {
+        return isAccessingSubfieldsInternally;
+    }
+
+    public Optional<Set<Integer>> getCallArgumentsToPassOuterLambdaSubfields()
+    {
+        return callArgumentsToPassOuterLambdaSubfields != null ? Optional.of(callArgumentsToPassOuterLambdaSubfields) : Optional.empty();
+    }
+
+    public boolean isTransformingInput()
+    {
+        return isTransformingInput;
+    }
+
+    public boolean isPruningSubfields()
+    {
+        return isPruningSubfields;
+    }
+
+    public List<LambdaDescriptor> getLambdaDescriptors()
+    {
+        return lambdaDescriptors;
+    }
+
+    public boolean isAcceptingLambdaArgument()
+    {
+        return !lambdaDescriptors.isEmpty();
+    }
+
+    public Optional<Function<Set<Subfield>, Optional<Set<Subfield>>>> getOuterLambdaSubfieldsTransform()
+    {
+        return outerLambdaSubfieldsTransform != null ? Optional.of(outerLambdaSubfieldsTransform) : Optional.empty();
+    }
+
+    /**
+     * Adds <code>allSubscripts</code> on top of the path for every subfield in 'subfields'.
+     * @param subfields set of Subfield to transform
+     * @return transformed copy of the input set of subfields with <code>allSubscripts</code>.
+     */
+    public static Optional<Set<Subfield>> prependAllSubscripts(Set<Subfield> subfields)
+    {
+        return Optional.of(subfields.stream().map(subfield -> new Subfield(subfield.getRootName(),
+                unmodifiableList(Stream.concat(Arrays.asList(allSubscripts()).stream(), subfield.getPath().stream()).collect(Collectors.toList()))))
+                .collect(Collectors.toSet()));
+    }
+
+    /**
+     * Removes the second path element from every subfield in 'subfields'.
+     * @param subfields set of Subfield to transform
+     * @return transformed copy of the input set of subfields with removed the second path element.
+     */
+    public static Optional<Set<Subfield>> removeSecondPathElement(Set<Subfield> subfields)
```

Comment:
This method looks weird to me? what is the use of this?
Anyways it is not being used so lets remove it?

Reply:
This method is used in `combination` function as 
```
outerLambdaSubfieldsTransform = {@SubfieldTransformationFunction(clazz = FunctionDescriptor.class, method = "removeSecondPathElement")},
```
the reason for it is that `combination` function transform the array into the array of arrays. 

Considering this table: 
```
CREATE TABLE documents(  words ARRAY(ROW(word VARCHAR, weight DOUBLE)) )
INSERT INTO documents VALUES ( ARRAY[ ROW('presto', 1.0), ROW('query', 0.9), ROW('engine', 0.9) ] )
```
For the query below
```
SELECT combinations(words) FROM documents
```
results will be 
```
ARRAY[ ARRAY[ ROW('presto', 1.0), ROW('query', 0.9) ], ARRAY[ ROW('query', 0.9), ROW('engine', 0.9) ], ARRAY[ ROW('engine', 0.9), ROW('presto', 1.0) ] ]
```
For the query
```
SELECT transform(combinations(words), x -> x[1].word) FROM documents
```
the captured subfield has path `words[*].[1].word`. 
Therefore, in order to correctly attribute the previously accessed subfields, we need to remove the second element  from the subfield path - `[1]` part.

I agree that probably this function is weird and is specific to `combination` function only. I can move it to the `ArrayCombinationsFunction` class

Reply:
Moved the function to `ArrayCombinationsFunction`